File: src/main/java/simulations/Pend.java
package simulations;

import java.awt.Graphics;
import java.util.ArrayList;
import java.awt.Color;
import java.util.List;

import spacesimulation.*;
import spacesimulation.algebra.*;
import spacesimulation.physics.Connection;
import spacesimulation.physics.Mass;

public class Pend extends MassSimulation {
    CartesianCoordinateSystem cart = new CartesianCoordinateSystem(true, 500000, 500000, Color.black);
    private final int maxRopeSegmentLength;
    private final int amountOfPoints;

    private final List<Connection> connections = new ArrayList<>();
    
    public Pend(int amountOfPoints, Simulator sim) {
        super(1, new Vec(0, -9.81, -0), sim);
        this.amountOfPoints = amountOfPoints;
        maxRopeSegmentLength = 100/amountOfPoints;
        reset();
        drawer.setZoom(3);
        drawer.setCameraAngleHorizontal(0.2);
    }

    @Override
    public void buffer() {
        for (int i = 1; i < masses.size(); i++) {
            if (masses.get(i - 1).getConnectingVectorTo(masses.get(i)).getLength() > maxRopeSegmentLength) {
                Vec posVec = masses.get(i - 1).getConnectingVectorTo(masses.get(i));
                double scalar = maxRopeSegmentLength / posVec.getLength();
                posVec.scale(scalar);
                masses.get(i).set(masses.get(i - 1).getPositionVector().add(posVec));
            }
        }
    }

    @Override
    public void calcForces() {
        getInput();
        for (Connection c : connections) {
            c.tick();
        }
    }

    private void getInput() {
        if (keymanager.f) masses.get(0).accelerate(new Vec(0.01, 0, 0));
        if (keymanager.g) masses.get(0).accelerate(new Vec(-0.01, 0, 0));
        if (keymanager.v) masses.get(0).accelerate(new Vec(0, 0, 5));
        if (keymanager.b) masses.get(0).accelerate(new Vec(0, 0, -5));
        if (keymanager.up) masses.get(1).applyForce(new Vec(10, 0, 0));
    }

    @Override
    public void render(Graphics g) {
        cart.render(drawer, g);
        for (Mass m : masses) m.render(drawer, g);
        for (Connection c : connections) c.render(drawer, g);

        for (int i = 0; i < masses.size(); i++) {
             g.drawString(masses.get(i).toString(), 10, 100 + i * 20);
        }
    }

    @Override
    public void reset() {
        masses.clear();
        for (int i = 0; i < amountOfPoints; i++) {
            addNewMass(new Point3d(0, -i * maxRopeSegmentLength * 0.7, -i * maxRopeSegmentLength * 0.7), i != 0);
        }
        masses.get(0).setStatus(Mass.Status.Immovable);

        connections.clear();
        for (int i = 0; i < amountOfPoints - 1; i++) {
            connections.add(new Connection(masses.get(i), masses.get(i + 1), maxRopeSegmentLength));
        }
    }
}

File: src/main/java/simulations/Pend.kt
package simulations

import Conf
import algebra.Point3d
import framework.WatchDouble
import framework.WatchInt
import framework.interfaces.Status
import framework.physics.ImpulseConnection
import framework.physics.PhysicsSimulation
import framework.physics.PointMass
import toVec
import kotlin.math.PI

@Suppress("unused")
class Pend(amountOfPoints: Int, length: Double) : PhysicsSimulation("String") {
    private val radius = 0.025
    private val links = ArrayList<ImpulseConnection>()
    private val masses = ArrayList<PointMass>()
    private var maxEnergy = 30.0
        set(value) = links.forEach { it.maxEnergy = value }
    private val maxRopeSegmentLength: Double
        get() = length / amountOfPoints
    @WatchDouble("Rope-Length", 1.0, 3.0)
    private var length: Double = length
        set(value) {
            field = value
            synchronized(links) {
                links.forEach { it.maxDistance = maxRopeSegmentLength }
            }
        }
    @WatchInt("Segments", 1, 100)
    private var amountOfPoints: Int = amountOfPoints
        set(value) {
            synchronized(links) {
                synchronized(masses) {
                    val delta = amountOfPoints - value
                    if (delta > 0) {
                        // field > value -> remove delta masses from the rope
                        repeat(delta) {
                            val removedMass = masses.removeLast()
                            val lastMass = masses.last()
                            val toBeRemoved = links.filter { it.isConnectedTo(removedMass) && it.isConnectedTo(lastMass) }
                            links.removeAll(toBeRemoved.toSet())
                            toBeRemoved.forEach { unregister(it) }
                        }
                    } else if (delta < 0) {
                        repeat(-delta) {
                            val lastMass = masses.last()
                            val mass = PointMass(
                                lastMass.mass,
                                lastMass.position.x,
                                lastMass.position.y,
                                lastMass.position.z - maxRopeSegmentLength * 0.8,
                                radius,
                            )
                            mass.color = Conf.colorScheme.smallObjectColor.toVec()
                            masses.add(mass)
                            register(mass)
                            val link = ImpulseConnection(lastMass, mass, maxRopeSegmentLength, maxEnergy)
                            links.add(link)
                            register(link)
                        }
                    }
                }
                links.forEach { it.maxDistance = maxRopeSegmentLength }
            }
            field = value
        }

    init {
        reset()
    }

    override fun render() {
        synchronized(masses) { for (m in masses) m.render(camera) }
        synchronized(links) { links.filter { !it.broken }.forEach { it.render(camera) } }
    }

    override fun calcForces() = Unit

    override fun reset() {
        super.reset()
        synchronized(masses) {
            masses.clear()
            for (i in 0 until amountOfPoints) {
                val pos = Point3d(-i * maxRopeSegmentLength, 0.0, 0.0)
                val mass = PointMass(1.0, pos.x, pos.y, pos.z, radius)
                if (i == 0) mass.status = Status.Immovable
                mass.color = Conf.colorScheme.smallObjectColor.toVec()
                masses.add(mass)
                register(mass)
            }
        }
        synchronized(links) {
            links.clear()
            for (i in 0 until amountOfPoints - 1) {
                val link = ImpulseConnection(masses[i], masses[i + 1], maxRopeSegmentLength, maxEnergy)
                links.add(link)
                register(link)
            }
        }
        camera.focalLength = 10.0
        camera.x = 0.0
        camera.y = -25.0
        camera.z = -length / 2
        camera.theta = PI / 2
        camera.phi = PI
        camera.focalLength = 10.0
        camera.zoom = 0.001
    }
}

fun main() {
    Pend(10, 2.0).start()
}